classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+namespace tool_mergeusers\task;
+defined('MOODLE_INTERNAL') || die();
+require_once(__DIR__ . '/../../lib/autoload.php');
+use MergeUserTool;
+use \tool_mergeusers\merge_request;
+/**
+ * Adhoc task to merge user accounts.
+ *
+ * @package   tool_mergeusers
+ * @author    Jordi Pujol-Ahulló <jordi.pujol@urv.cat>
+ * @author    Nicola Vallinoto <n.vallinoto@liguriadigitale.it>
+ * @copyright 2013 Servei de Recursos Educatius (http://www.sre.urv.cat)
+ * @copyright 2023 Liguria Digitale (http://www.liguriadigitale.it)
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * */
+class merge_user_accounts extends \core\task\adhoc_task {
+    /**
+     * Execute the task of merge users accounts.
+     */
+    public function execute() {
+        global $DB;
+        $data = $this->get_custom_data();
+        $maxattempts = get_config('tool_mergeusers', 'maxattempts');
+        $mergerequest = $DB->get_record(merge_request::TABLE_MERGE_REQUEST,
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
```

Comment:
I would use the merge_request class to make it simple to manage all the stuff:

```php
$mergerequest = merge_request::from($mergerequest);
```

And then, use it as it is. In particular, you will not have to deal with logs conversion nor anything unnecessary on the client side. Use the new proposed `append_log()` instead of dealing with arrays and so. The rest of fields can be set and get as always, thanks to the magic `__set` and `__get` methods.

Reply:
Then, it will be easier also to pass a `mergerequest $mergerequest` instance to a method, than simply an `object $mergerequest`.

Reply:
it's not clear to me as proceed. first of all how to change (you told to make a stdclass - please indicate me another stdclass inside moodle or outside if any) and to put (in which folder) the actual merge_request.php

Reply:
> 

when you suggest to add 
`$mergerequest = merge_request::from($mergerequest);` ?

after this lines ?
`if (!$mergerequest) {
            // Merge request was removed before running this task.
            return;
        }`

in the case the execution will give me an error:

Adhoc task failed: tool_mergeusers\task\merge_user_accounts,tool_mergeusers\merge_request::from(): Argument #1 ($data) must be of type tool_mergeusers\stdClass, stdClass given, called in /srv/http/moodlesvi.liguriadigitale.it/admin/tool/mergeusers/classes/task/merge_user_accounts.php on line 45

Reply:
You have the answer on the error message.

You need to add `use stdClass;` on the `tool_mergeusers` class file.

Reply:
it means to change this code line inside merge_request.php
`private $data;`  
in 
`private strClass $data;` 
?

Reply:
ok added